headless/find.go

```diff
@@ -0,0 +1,225 @@
+	Message string `json:"message,omitempty"`
+}
+
+func errorResp(c *gin.Context, error_code int, err error) {
+	c.JSON(error_code, FindRespond{
+		Message: err.Error(),
+	})
+}
+
+func validate(c *gin.Context) {
+	var req FindRequest
+	if err := c.Bind(&req); err != nil {
+		errorResp(c, http.StatusBadRequest, xerrors.Errorf("Param error"))
+		return
+	}
+
+	if err := checkValidateRequest(&req); err != nil {
+		errorResp(c, http.StatusBadRequest, err)
+		return
+	}
+
+	launcher := newLauncher(LauncherPath)
+	defer launcher.Cleanup()
+	defer launcher.Kill()
+
+	u, err := launcher.Launch()
+	if err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+
+	browser := rod.New().ControlURL(u)
+	if err := browser.Connect(); err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+
+	defer browser.Close()
+
+	page, err := browser.Page(proto.TargetCreateTarget{URL: req.Location})
+	if err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+
+	timeout := req.Timeout
+	if timeout == "" {
+		timeout = defaultTimeout
+	}
+
+	timeoutDuration, err := time.ParseDuration(timeout)
+	if err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+	router := page.HijackRequests()
+
+	resources := []proto.NetworkResourceType{
+		proto.NetworkResourceTypeFont,
+		proto.NetworkResourceTypeImage,
+		proto.NetworkResourceTypeMedia,
+		proto.NetworkResourceTypeStylesheet,
+		proto.NetworkResourceTypeWebSocket, // we don't need websockets to fetch html
+	}
+
+	router.MustAdd("*", func(ctx *rod.Hijack) {
+		if slice.Contains(resources, ctx.Request.Type()) {
+			ctx.Response.Fail(proto.NetworkErrorReasonBlockedByClient)
+			return
+		}
+
+		ctx.ContinueRequest(&proto.FetchContinueRequest{})
+	})
+
+	go router.Run()
+	defer router.Stop()
+
+	page = page.Timeout(timeoutDuration)
+	if err := page.WaitLoad(); err != nil {
+		errorResp(c, http.StatusInternalServerError, xerrors.Errorf("%w", err))
+		return
+	}
+
+	// Wait for XHR
+	page.WaitNavigation(proto.PageLifecycleEventNameNetworkAlmostIdle)()
+
+	switch req.Match.Type {
+	case matchTypeRegex:
+		selector := req.Match.MatchRegExp.Selector
+		if selector == "" {
+			selector = "*"
+		}
+
+		if _, err := page.ElementR(selector, req.Match.MatchRegExp.Value); err != nil {
+			c.JSON(http.StatusOK, FindRespond{Found: false, Message: err.Error()})
+
+			return
+		}
+	case matchTypeXPath:
+		selector := req.Match.MatchXPath.Selector
+		if _, err := page.ElementX(selector); err != nil {
+			c.JSON(http.StatusOK, FindRespond{Found: false, Message: err.Error()})
+
+			return
+		}
+	case matchTypeJS:
+		js := req.Match.MatchJS.Value
+		if _, err := page.ElementByJS(rod.Eval(js)); err != nil {
+			c.JSON(http.StatusOK, FindRespond{Found: false, Message: err.Error()})
+
+			return
+		}
+	}
+
+	c.JSON(http.StatusOK, FindRespond{Found: true})
```

Comment:
We need to get the full content of query result in response.

for example, here's a request:

```json
{
  "location": "https://www.minds.com/newsfeed/1421043369127186449",
  "timeout": "30s",
  "match": {
     "type": "regexp",
     "regexp": {
       "selector": "*",
       "value": "Sig"
     }
  }
}
```

This service will respond with `found: true`, which is pretty good. 
But what upstream indeed cares is the whole post content (contents after `Sig:` in this post).
So we need to return it in API response.

Reply:
Or, at least, if I define `type: regexp` with search condition: `^Sig: .*$`, the whole matched substring should be returned in API response.
Same for `xpath` and `js` selector: the node's content should be returned. Then `proof_service` upstream will have a chance to verify the signature in it.

Reply:
>  the returned value from this API should be consistent for all match types.

Agree.

>  it might return the node's content regardless of what match type.

Yeah this is what I think. Even "kinda dirty" match result [1] returned can be tolerated, because all `proof_service` upstream wants is basiclly only one line: `Sig: BASE64_SIGNATURE` .

[1]: like some `<div>` tag is mixed into the returned result.